skbio/io/format/genbank.py

```diff
@@ -0,0 +1,487 @@
+import re
+import textwrap
+
+import numpy as np
+import pandas as pd
+from datetime import datetime
+from functools import partial
+
+from skbio.io import create_format, GenbankFormatError
+from skbio.io.registry import FileSentinel
+from skbio.io.format._base import (_get_nth_sequence,
+                                   _parse_fasta_like_header,
+                                   _format_fasta_like_records, _line_generator,
+                                   _too_many_blanks)
+from skbio.util._misc import chunk_str
+from skbio.sequence import Sequence, DNA, RNA, Protein
+from pprint import pprint
+
+genbank = create_format('genbank')
+
+# date format in locus line of genbank record
+_TIME_FORMAT = '%d-%b-%Y'
+# This list is ordered
+# used to read and write genbank file.
+_HEADERS = ['LOCUS',
+            'DEFINITION',
+            'ACCESSION',
+            'VERSION',
+            'DBSOURCE',
+            'DBLINK',
+            'KEYWORDS',
+            'SOURCE',
+            'REFERENCE',
+            'COMMENT',
+            'FEATURES',
+            'ORIGIN']
+
+
+@genbank.sniffer()
+def _genbank_sniffer(fh):
+    # check the 1st real line is a valid LOCUS line
+    if _too_many_blanks(fh, 5):
+        return False, {}
+    try:
+        line = next(_line_generator(fh, skip_blanks=True, strip=False))
+    except StopIteration:
+        return False, {}
+
+    try:
+        _parse_locus([line])
+    except GenbankFormatError:
+        return False, {}
+    return True, {}
+
+
+@genbank.reader(None)
+def _genbank_to_generator(fh, constructor=None, **kwargs):
+    for record in _parse_genbanks(fh):
+        yield _construct(record, constructor, **kwargs)
+
+
+@genbank.reader(Sequence)
+def _genbank_to_biological_sequence(fh, seq_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), seq_num)
+    return _construct(record, Sequence, **kwargs)
+
+
+@genbank.reader(DNA)
+def _genbank_to_dna(fh, seq_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), seq_num)
+    return _construct(record, DNA, **kwargs)
+
+
+@genbank.reader(RNA)
+def _genbank_to_rna(fh, seq_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), seq_num)
+    return _construct(record, RNA, **kwargs)
+
+
+@genbank.reader(Protein)
+def _genbank_to_protein(fh, seq_num=1, **kwargs):
+    record = _get_nth_sequence(_parse_genbanks(fh), seq_num)
+    return _construct(record, Protein, **kwargs)
+
+
+@genbank.writer(None)
+def _sequences_to_genbank(fh):
+    pass
+
+
+def _construct(record, constructor=None, **kwargs):
+    seq, md, pmd = record
+    if constructor is None:
+        unit = md['LOCUS']['unit']
+        if unit == 'bp':
+            # RNA mol type has T instead of U;
+            # so still read in as DNA
+            constructor = DNA
+        elif unit == 'aa':
+            constructor = Protein
+        else:
+            constructor = Sequence
+    if 'lowcase' not in kwargs and constructor != Sequence:
```

Comment:
sorry, I had realized it and already have it in one of the following commit
that just got pushed a sec ago...

On Tue, Aug 11, 2015 at 3:09 PM, Evan Bolyen <notifications@github.com>
wrote:

> In skbio/io/format/genbank.py
> <https://github.com/biocore/scikit-bio/pull/1023#discussion_r36805596>:
>
> > +    pass
> > +
> > +
> > +def _construct(record, constructor=None, **kwargs):
> > +    seq, md, pmd = record
> > +    if constructor is None:
> > +        unit = md['LOCUS']['unit']
> > +        if unit == 'bp':
> > +            # RNA mol type has T instead of U;
> > +            # so still read in as DNA
> > +            constructor = DNA
> > +        elif unit == 'aa':
> > +            constructor = Protein
> > +        else:
> > +            constructor = Sequence
> > +    if 'lowcase' not in kwargs and constructor != Sequence:
>
> typo: 'lowcase'
>
> Also, I have a PR open: #1043
> <https://github.com/biocore/scikit-bio/pull/1043> which fixes this
> generally.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/scikit-bio/pull/1023/files#r36805596>.
>